tianocore/edk2
# Bug 4683 unit test unexpected exception may be seen as PASS condition

REF: https://bugzilla.tianocore.org/show_bug.cgi?id=4683

The main bug is that EDK II CI can miss a unit test failure when
a unit test generates an exception. Addressing this one issue 
required the following chain of changes to the UnitTestFrameworkPkg
that are included in this patch series.

New in v4
==========
* Moved GoogleTestLib.inf /EHsc --> EHs from Patch 2 ->Patch 3
  so the /EHsc change is not mixed with /MT changes in Patch 2.
* Fixed typo in Patch 1 commit message

New in V3
=========
* Add EXPECT_THROW_MESSAGE() and ASSERT_THROW_MESSAGE() to 
  GoogleTestLib.h to simplify unit test case statements testing for 
  an ASSERT() for a specific trigger expressiom. Uncrustify format
  of EXPECT_THAT() with [](){stmt} make code harder to read. Adding
  macros makes the source content in unit test cases match the 
  source style of EXPECT_THROW() statements.
* Update GoogTest samples to use EXPECT_THROW_MESSAGE() instead of 
  EXPECT_THAT().

New in V2
=========
* Include Throws() APIs in GoogleTestLib.h
* Update GoogleTest samples to demonstrate use of EXPECT_THROW()
  and EXPECT_THAT() to test for more specific ASSERT() conditions
  including the ability to verify the specific ASSERT() expression
  that was triggered.

V1
=========
* For host-base VX20xx builds, use release builds by default to
  make sure unexpected exceptions are caught as a test failure.
  Using debug libraries may generate a popup window to prompt 
  running a debugger, which is value for debug, but that mode
  does not generate an error in CI.
* Update UNIT_TEST_DEBUG to enable use of VS20xx debug libraries
  to enable popup windows to run debugger. 
* Remove DebugLib.h internal macro name collision with windows.h
  that was introduced when release libraries are used by default.
* Update MSFT and GCC flags to consistently enable structured 
  exception handling so Google Test host-based unit tests that
  check for expected ASSERT() conditions can be properly detected.
  The additional benefit of this bug fix is that catching thrown
  exceptions is much faster than EXPECT_DEATH() tests. The gtest
  sample unit tests are updated to use ASSERT_ANY_THROW() instead
  of EXPECT_DEATH().
* In order to catch C++ exceptions for google test, the 
  UnitTestDebugAssertLib has to be split into a target version
  and a host version and had to be able to detect the difference
  between cmock host-based tests and gtest host-based tests. This
  required minor changes to the UnitTestLib to be able to use the
  UnitTestLib GetFrameworkHandle() API to make that determination.
  It is NULL for gtest and non-NULL for cmocka.
* In order to test the unit test failing conditions, unit tests
  the fail on purpose and unit tests that generate exceptions are
  added.  These are not added to CI because that would cause CI to
  always fail. Instead, they can be run to make sure that reports
  generated provide the right level of detail for a developer to
  quickly identify the source of the unit test failure or the 
  source of the exception.

Cc: Michael Kubacki [mikuback@linux.microsoft.com](mailto:mikuback@linux.microsoft.com)
Cc: Sean Brogan [sean.brogan@microsoft.com](mailto:sean.brogan@microsoft.com)
Cc: Liming Gao [gaoliming@byosoft.com.cn](mailto:gaoliming@byosoft.com.cn)
Cc: Zhiguang Liu [zhiguang.liu@intel.com](mailto:zhiguang.liu@intel.com)
Cc: Andrew Fish [afish@apple.com](mailto:afish@apple.com)
Cc: Leif Lindholm [quic_llindhol@quicinc.com](mailto:quic_llindhol@quicinc.com)
Signed-off-by: Michael D Kinney [michael.d.kinney@intel.com](mailto:michael.d.kinney@intel.com)
